### PR TITLE
Feature/double precision

### LINKED ISSUE #2497: Add Float64 support
Float64 precision is currently not supported. 
Why? This should be pretty straightforward, raising an Exception if the device or other configurations are not compatible with it.
Are you planning to add support soon? 
What is the best workaround currently? Will it work if I do it manually via `model.to(torch.float64)` and similar for each dataloader or are there some caveats?

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #2497 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/precision/double.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    def connect(
+        self,
+        model: 'Module',
+        optimizers: Sequence['Optimizer'],
+        lr_schedulers: Sequence[Any],
+    ) -> Tuple['Module', Sequence['Optimizer'], Sequence[Any]]:
+        """Converts the model to double precision and wraps the forward to convert incoming floating point data.
+        Does not alter `optimizers` or `lr_schedulers`."""
+        def patch_forward(old_forward: Callable) -> Callable:
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/double.py:45]: can you move this as a static function to the class? So that it can be pickled?

File: pytorch_lightning/plugins/precision/double.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            def new_forward(*args, **kwargs) -> Any:
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/double.py:46]:
```suggestion
            @wraps(old_forward)
            def new_forward(*args, **kwargs) -> Any:
```

File: pytorch_lightning/plugins/precision/double.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                return old_forward(*apply_to_collection(args, torch.Tensor, _to_double_precision),
+                                   **apply_to_collection(kwargs, torch.Tensor, _to_double_precision))
+            return new_forward
+        model = model.to(dtype=torch.float64)
+        model.forward = patch_forward(model.forward)
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/double.py:51]: So the model's forward will be patched forever? Can we maybe revert this once we finished training/testing/predicting?

File: tests/plugins/test_double_plugin.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import torch
+
+from pytorch_lightning import Trainer
+from tests.helpers import BoringModel
+
+
+class DoublePrecisionBoringModel(BoringModel):
+
+    def training_step(self, batch, batch_idx):
+        output = self((batch, torch.ones_like(batch).long()))  # Add some non floating-point data
```
Comment (justusschock) [inline tests/plugins/test_double_plugin.py:10]: can you also assert that here the data is of fp32? Just to make sure that we really convert and not our tests just pass because we passed already converted data :)

**Review by justusschock — COMMENTED** [review]
Hi @ethanwharris ,

Great to see you again :)

And great PR!
Just some smaller comments :)

**Comment by ethanwharris** [comment]
Hi @justusschock,

Thanks for the great suggestions, all done :)

**Comment by ethanwharris** [comment]
The teardown is called at the end of training, but won't be called before checkpointing. Any way to address that?

**Comment by williamFalcon** [comment]
omg.... i love this feature 
😍😍😍😍😍

**Comment by rohitgr7** [comment]
can we add support with deepspeed too?
https://github.com/PyTorchLightning/pytorch-lightning/blob/3a56a6024e0d8b239801cce558381807c24ba3d0/pytorch_lightning/plugins/training_type/deepspeed.py#L46-L48

not sure why it's handled this way in here, not an expert with deepspeed.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -319,7 +320,8 @@ def select_precision_plugin(self) -> PrecisionPlugin:
 
         if self.precision == 32:
             return PrecisionPlugin()
-
+        elif self.precision == 64:
+            return DoublePrecisionPlugin()
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:324]: also update the error on line 363

Comment (ethanwharris) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:324]: Good catch! Done :)

File: pytorch_lightning/plugins/precision/double.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    def _teardown(model: 'Module', old_forward: Callable) -> None:
+        model.forward = old_forward
+
+    def _patch_forward(self, model: 'Module') -> None:
+        old_forward = model.forward
+
+        @wraps(old_forward)
+        def new_forward(*args, **kwargs) -> Any:
+            return old_forward(*apply_to_collection(args, torch.Tensor, DoublePrecisionPlugin._to_double_precision),
+                               **apply_to_collection(kwargs, torch.Tensor, DoublePrecisionPlugin._to_double_precision))
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/double.py:48]: You can apply the same pattern as here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/3a56a6024e0d8b239801cce558381807c24ba3d0/pytorch_lightning/plugins/training_type/deepspeed.py#L52-L58

By checking for `torch.DoubleTensor`

Comment (ethanwharris) [reply pytorch_lightning/plugins/precision/double.py:48]: The method I used allowed for filtering using `torch.is_floating_point`, which is more reliable than just type checking for `FloatTensor` and `cuda.FloatTensor`. I can change this if needed but I think `is_floating_point` is a better approach. I modified this from the source for Nvidia Apex, which seems quite reliable :)

**Comment by SeanNaren** [comment]
Thanks for the awesome contribution @ethanwharris!

Out of curiosity, if the user doesn't use the forward function of the LightningModule and instead calls independent modules, does it still work? i.e if I define my `training_step` like:

```python
 def training_step(self, batch, batch_idx):
        output = self.layer((batch, torch.ones_like(batch).long())) # use layer instead of self(...)
```

**Review by tchaton — APPROVED** [review]
Nice !

**Comment by ethanwharris** [comment]
@SeanNaren That's an excellent point haha.

@SeanNaren @carmocca @justusschock I've made some changes to address your comments:
- Moved patch logic to seperate class that now handles teardown (much cleaner than before)
- Added patch to training_step etc. methods so that it all still works if the model is called directly without forward (and added a simple test of this behaviour)

@rohitgr7 Not sure about deepspeed, don't feel comfortable trying to add it as I don't currently have a way of testing it out. It seems that at least DeepSpeed ZeRO requires `precision=16`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/853523ee643fe0f0cc30d40d9e85a8869e7edfd8/tests/plugins/test_deepspeed_plugin.py#L193

Should all be ready for review now :smiley:

File: tests/plugins/test_double_plugin.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+def test_double_precision(tmpdir, boring_model):
+    model = boring_model()
+    original_training_step = model.training_step
+
+    trainer = Trainer(
+        max_epochs=2,
+        default_root_dir=tmpdir,
+        limit_train_batches=2,
+        limit_test_batches=2,
+        limit_val_batches=2,
```
Comment (carmocca) [inline tests/plugins/test_double_plugin.py:109]:
```suggestion
        fast_dev_run=2,
```

File: tests/plugins/test_double_plugin.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        assert output.dtype == torch.float64
+        return output
+
+    def predict_dataloader(self):
+        return DataLoader(RandomDataset(32, 64))
+
+
+@pytest.mark.parametrize(
+    ['boring_model'],
+    [(DoublePrecisionBoringModel,), (DoublePrecisionBoringModelNoForward,)]
```
Comment (carmocca) [inline tests/plugins/test_double_plugin.py:98]:
```suggestion
    'boring_model',
    (DoublePrecisionBoringModel, DoublePrecisionBoringModelNoForward)
```

File: pytorch_lightning/plugins/precision/double.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+class _DoublePrecisionPatch:
+    """Class to handle patching of methods in the `LightningModule` and subsequent teardown.`"""
+
+    def __init__(self, model: 'Module', method_name: str, old_method: Any):
+        self.model = model
+        self.method_name = method_name
+        self.old_method = old_method
+
+    def teardown(self):
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/double.py:36]:
```suggestion
    def teardown(self) -> None:
```

File: pytorch_lightning/plugins/precision/double.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+if TYPE_CHECKING:
+    from torch.nn import Module
+    from torch.optim import Optimizer
+
+
+class _DoublePrecisionPatch:
+    """Class to handle patching of methods in the `LightningModule` and subsequent teardown.`"""
+
+    def __init__(self, model: 'Module', method_name: str, old_method: Any):
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/double.py:31]:
```suggestion
    def __init__(self, model: 'Module', method_name: str, old_method: Any) -> None:
```

File: pytorch_lightning/plugins/precision/double.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        return data
+
+    @staticmethod
+    def _move_float_tensors_to_double(collection: Any):
+        return apply_to_collection(
+            collection, torch.Tensor, function=_DoublePrecisionPatch._to_double_precision
+        )
+
+    @staticmethod
+    def patch(model: 'Module', method_name: str):
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/double.py:52]:
```suggestion
    def patch(model: 'Module', method_name: str) -> _DoublePrecisionPatch:
```

File: pytorch_lightning/plugins/precision/double.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        setattr(model, method_name, new_method)
+        return _DoublePrecisionPatch(model, method_name, old_method)
+
+
+class DoublePrecisionPlugin(PrecisionPlugin):
+    """Plugin for training with double (`torch.float64`) precision."""
+
+    precision: int = 64
+
+    def __init__(self):
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/double.py:71]:
```suggestion
    def __init__(self) -> None:
```

File: tests/helpers/boring_model.py [hunk 13]
```diff
@@ -47,6 +47,20 @@ def __len__(self):
         return self.len
 
 
+class RandomFloatIntDataset(Dataset):
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply tests/helpers/boring_model.py]: No problem, all done :)

File: pytorch_lightning/plugins/precision/double.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        old_method = getattr(model, method_name)
+
+        @wraps(old_method)
+        def new_method(*args, **kwargs) -> Any:
+            return old_method(
+                *_DoublePrecisionPatch._move_float_tensors_to_double(args),
+                **_DoublePrecisionPatch._move_float_tensors_to_double(kwargs)
+            )
+
+        setattr(model, method_name, new_method)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/double.py:62]:
```suggestion
        setattr(model, method_name, new_method if callable(old_method) else old_method)
```
just to avoid cases if some method is set to None, although not a good practice.

File: pytorch_lightning/plugins/precision/double.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        return _DoublePrecisionPatch(model, method_name, old_method)
+
+
+class DoublePrecisionPlugin(PrecisionPlugin):
+    """Plugin for training with double (`torch.float64`) precision."""
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/double.py:67]:
```suggestion
    """Plugin for training with double (``torch.float64``) precision."""
```

File: pytorch_lightning/plugins/precision/double.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+from pytorch_lightning.plugins.precision.precision_plugin import PrecisionPlugin
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+
+if TYPE_CHECKING:
+    from torch.nn import Module
+    from torch.optim import Optimizer
+
+
+class _DoublePrecisionPatch:
+    """Class to handle patching of methods in the `LightningModule` and subsequent teardown.`"""
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/double.py:29]:
```suggestion
    """Class to handle patching of methods in the ``LightningModule`` and subsequent teardown."""
```

**Review by carmocca — APPROVED** [review]
LGTM. Can you have a look at the failing tests?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/precision/double.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+        return data
+
+    @staticmethod
+    def _move_float_tensors_to_double(collection: Any) -> Any:
+        return apply_to_collection(
+            collection, torch.Tensor, function=_DoublePrecisionPatch._to_double_precision
+        )
+
+    @staticmethod
+    def patch(model: 'Module', method_name: str) -> '_DoublePrecisionPatch':
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/double.py:52]:
```suggestion
    @classmethod
    def patch(cls, model: 'Module', method_name: str) -> '_DoublePrecisionPatch':
```

File: pytorch_lightning/plugins/precision/double.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+        @wraps(old_method)
+        def new_method(*args, **kwargs) -> Any:
+            return old_method(
+                *_DoublePrecisionPatch._move_float_tensors_to_double(args),
+                **_DoublePrecisionPatch._move_float_tensors_to_double(kwargs)
+            )
+
+        setattr(model, method_name, new_method if callable(old_method) else old_method)
+        return _DoublePrecisionPatch(model, method_name, old_method)
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/double.py:63]:
```suggestion
        return cls(model, method_name, old_method)
```

**Review by justusschock — APPROVED** [review]
2 more minor comments, which allow subclassing (and don't hardcode the class name, else LGTM

**Comment by ethanwharris** [comment]
@carmocca I have finally fixed the tests. `trainer.predict` calls `torch.set_grad_enabled(False)` which was then killing gradients for subsequent tests. I've added a call to `torch.set_grad_enable(True)` which fixes this for now. I'll open an issue to change the predict behaviour to re-enable grads, maybe just switch to `with torch.no_grad`.

**Review by rohitgr7 — APPROVED** [review]
